### PR TITLE
Improve APIResponseException class

### LINKED ISSUE #44: Add 4XX errors body to exception
Usually errors 4XX have useful explanation about an occurred error, but the library does parse and return error body only if error code is < 400.

See `ApiResource` line 194.

I think it's better to do something like
```php
  // Handle 4XX responses, 5XX responses will throw an HttpAdapterException
if ($statusCode < 500) {
    $body = json_decode($response->getBody()->getContents(), true);
    if ($statusCode < 400) {
        return $body;
    } elseif ($statusCode === 404) {
      throw new APIResponseException('The specified resource does not exist', 404);
    } else {
        throw new APIResponseException('Received bad response from ' . ucfirst($this->endpoint) . ' API: '. $statusCode.', body:' . print_r($body));
    }
}
```

I'm ready to prepare a PR.

### PR DESCRIPTION
**Pull request by zaporylie**
Hi

I work on SparkPost module for Drupal. I had some troubles sending email and Exception message and Code doesn't really help to investigate what was the source of the problem. API response has quite nice error message handling so I've added arguments to APIResponseException definition to aggregate API error messages.

`__construct($message = "", $code = 0, $apiMessage = "", $apiCode = 0, $apiMessageDescription = "")`

where

`$message` is custom message, 
`$code` is HTTP status Code,
`$apiMessage` is API error message received from endpoint,
`$apiCode` is API error code received from endpoint, 
`$apiMessageDescription` is API error message description received from endpoint.

Nothing special but it should improve DX, fix #44 and is backward compatible.

### CODE REVIEW

**Comment by chandon** [comment]
+1

File: lib/SparkPost/APIResponseException.php [hunk 1]
```diff
@@ -3,7 +3,53 @@
...
+   */
+  public function getAPICode() {
+    return $this->apiCode;
+  }
+
+  /**
+   * Gets the Exception message
+   * @return string the Exception message as a string.
+   */
+  public function getAPIMessageDescription() {
```
Comment (richleland) [inline lib/SparkPost/APIResponseException.php]: For consistency with the field names from the API, what do you think about calling this method `getAPIDescription`?

**Comment by richleland** [comment]
Hi @zaporylie this PR looks great - can we get the methods on the exception class covered? https://coveralls.io/builds/5351334/source?filename=lib%2FSparkPost%2FAPIResponseException.php

**Comment by zaporylie** [comment]
Renamed `getAPIMessageDescription` to `getAPIDescription`, added `APIResourceExceptionTest`

**Comment by richleland** [comment]
Thanks for the contribution @zaporylie.

**Pull request merged**